core/src/main/java/org/apache/iceberg/ManifestFilterManager.java

```diff
@@ -360,23 +360,23 @@ private boolean manifestHasDeletedFiles(PartitionAndMetricsEvaluator evaluator,
           dropPartitions.contains(file.specId(), file.partition()) ||
           (isDelete && entry.sequenceNumber() > 0 && entry.sequenceNumber() < minSequenceNumber);
 
-      boolean nonMatchingDeleteFile = !file.content().equals(FileContent.DATA) && !evaluator.rowsMustMatch(file);
-      if (!markedForDelete && nonMatchingDeleteFile) {
-        // not all DeleteFiles removal can be handled by metadata operation, skip in this case
-        continue;
-      }
-
       if (markedForDelete || evaluator.rowsMightMatch(file)) {
+        boolean allRowsMatch = markedForDelete || evaluator.rowsMustMatch(file);
         ValidationException.check(
-            markedForDelete || evaluator.rowsMustMatch(file),
+            allRowsMatch || isDelete, // ignore delete files where some records may not match the expression
             "Cannot delete file where some, but not all, rows match filter %s: %s",
             this.deleteExpression, file.path());
 
-        hasDeletedFiles = true;
+        hasDeletedFiles = allRowsMatch;
+
         if (failAnyDelete) {
           throw new DeleteException(reader.spec().partitionToPath(file.partition()));
```

Comment:
In failAnyDelete case, this would  trigger if we have a deleteFile but rows do not match.  Probably this never happens in delete call, but wonder if we need to update the condition (failAnyDelete && allRowsMatch).

Reply:
Good catch. I'll fix.